go-controller/pkg/ovn/base_network_controller_secondary.go

```diff
@@ -282,3 +308,154 @@ func (bsnc *BaseSecondaryNetworkController) syncPodsForSecondaryNetwork(pods []i
+		}
+		bsnc.namespaces[ns] = nsInfo
+	} else {
+		nsInfoExisted = true
+		// if we found an existing nsInfo, do not hold the namespaces lock
+		// while waiting for nsInfo to Lock
+		bsnc.namespacesMutex.Unlock()
+	}
+
+	var unlockFunc func()
+	if readOnly {
+		unlockFunc = func() { nsInfo.RUnlock() }
+		nsInfo.RLock()
+	} else {
+		unlockFunc = func() { nsInfo.Unlock() }
+		nsInfo.Lock()
+	}
+
+	if nsInfoExisted {
+		// Check that the namespace wasn't deleted while we were waiting for the lock
+		bsnc.namespacesMutex.Lock()
+		defer bsnc.namespacesMutex.Unlock()
+		if nsInfo != bsnc.namespaces[ns] {
+			unlockFunc()
+			return nil, nil, fmt.Errorf("namespace %s, was removed during ensure", ns)
+		}
+	}
+
+	// nsInfo and namespace didn't exist, get it from lister
+	if namespace == nil {
+		var err error
+		namespace, err = bsnc.watchFactory.GetNamespace(ns)
+		if err != nil {
+			namespace, err = bsnc.client.CoreV1().Namespaces().Get(context.TODO(), ns, metav1.GetOptions{})
+			if err != nil {
+				klog.Warningf("Unable to find namespace during ensure in informer cache or kube api server. " +
+					"Will defer configuring namespace.")
+			}
+		}
+	}
+
+	if namespace != nil {
+		// if we have the namespace, attempt to configure nsInfo with it
+		if err := bsnc.configureNamespaceCommon(nsInfo, namespace); err != nil {
+			unlockFunc()
+			return nil, nil, fmt.Errorf("failed to configure namespace %s: %v", ns, err)
+		}
+	}
+
+	return nsInfo, unlockFunc, nil
+}
+
+func (bsnc *BaseSecondaryNetworkController) updateNamespaceForSecondaryNetwork(old, newer *kapi.Namespace) error {
+	var errors []error
+	klog.Infof("[%s] updating namespace for network %s", old.Name, bsnc.GetNetworkName())
+
+	nsInfo, nsUnlock := bsnc.getNamespaceLocked(old.Name, false)
+	if nsInfo == nil {
+		klog.Warningf("Update event for unknown namespace %q", old.Name)
+		return nil
+	}
+	defer nsUnlock()
+
+	aclAnnotation := newer.Annotations[util.AclLoggingAnnotation]
+	oldACLAnnotation := old.Annotations[util.AclLoggingAnnotation]
+	// support for ACL logging update, if new annotation is empty, make sure we propagate new setting
+	if aclAnnotation != oldACLAnnotation {
+		if err := bsnc.updateNamespaceAclLogging(old.Name, aclAnnotation, nsInfo); err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	if err := bsnc.multicastUpdateNamespace(newer, nsInfo); err != nil {
+		errors = append(errors, err)
+	}
+	return kerrors.NewAggregate(errors)
+}
+
+func (bsnc *BaseSecondaryNetworkController) deleteNamespace4SecondaryNetwork(ns *kapi.Namespace) error {
+	klog.Infof("[%s] deleting namespace for network %s", ns.Name, bsnc.GetNetworkName())
+
+	nsInfo := bsnc.deleteNamespaceLocked(ns.Name)
+	if nsInfo == nil {
+		return nil
+	}
+	defer nsInfo.Unlock()
+
+	if err := bsnc.multicastDeleteNamespace(ns, nsInfo); err != nil {
+		return fmt.Errorf("failed to delete multicast nameosace error %v", err)
+	}
+	return nil
+}
```

Comment:
honestly, I am not sure what is the better way to share the code. Especially between ensureNamespaceLocked() and ensureNamespaceLockedForSecondaryNetwork()

Reply:
That's one option. 

The other option is to split ensureNamespaceLocked because the first 60 lines of that method are practically the same than ensureNamespaceLockedForSecondaryNetwork.

Reply:
I don't see how to split it based on how locking is done today.

Reply:
> what will we be doing differently between the default network controller and the secondary network controller with IPAM?
> 
> The way I see it, the big difference will come for an IPAMless secondary controller, which will not have to track namespaces / pods labels and keep IP based address sets in sync.

then for ipamless secondary controller, we do not need watch for namespaces at all. I've made the change.